cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java

```diff
@@ -0,0 +1,474 @@
+    // assumption is there is only a single runnable
+    for (Map.Entry<String, String> envEntry : environments.values().iterator().next().entrySet()) {
+      V1EnvVar envVar = new V1EnvVar();
+      envVar.setName(envEntry.getKey());
+      envVar.setValue(envEntry.getValue());
+      envVars.add(envVar);
+    }
+    container.setEnv(envVars);
+    V1ResourceRequirements resourceRequirements = new V1ResourceRequirements();
+    Map<String, Quantity> quantityMap = new HashMap<>();
+    quantityMap.put("cpu", new Quantity(String.valueOf(vcores)));
+    quantityMap.put("memory", new Quantity(String.format("%dMi", memoryMB)));
+    resourceRequirements.setRequests(quantityMap);
+    container.setResources(resourceRequirements);
+    // when other program types are supported, there will need to be a mapping from program type to main class
+    container.setArgs(Arrays.asList("co.cask.cdap.internal.app.runtime.k8s.UserServiceProgramMain",
+                                    "--env=k8s", String.format("--appSpecPath=%s/%s", podInfo.getDir(), APP_SPEC),
+                                    String.format("--programOptions=%s/%s", podInfo.getDir(), PROGRAM_OPTIONS)));
+    podSpec.setContainers(Collections.singletonList(container));
+
+    templateSpec.setSpec(podSpec);
+    spec.setTemplate(templateSpec);
+    deployment.setSpec(spec);
+
+    LOG.trace("Creating deployment {} in Kubernetes", resourceMeta.getName());
+    deployment = appsApi.createNamespacedDeployment(kubeNamespace, deployment, "true");
+    LOG.trace("Creating deployment {} in Kubernetes", resourceMeta.getName());
+    return deployment;
+  }
+
+  /**
+   * Create the config map for the program run. Since the config map contains the program options, which contains
+   * runtime arguments, it needs to be created for each program run and deleted when the run ends.
+   */
+  private V1ConfigMap createConfigMap() throws ApiException, IOException {
+    CoreV1Api api = new CoreV1Api(apiClient);
+    V1ConfigMap configMap = new V1ConfigMap();
+    configMap.setMetadata(resourceMeta);
+    configMap.putBinaryDataItem(APP_SPEC, Files.readAllBytes(new File(appSpec).toPath()));
+    configMap.putBinaryDataItem(PROGRAM_OPTIONS, Files.readAllBytes(new File(programOptions).toPath()));
+    // also add the pod.labels.properties file used by Kubernetes service creation
+    String labelsStr = resourceMeta.getLabels().entrySet().stream()
+      .map(entry -> String.format("%s=\"%s\"", entry.getKey(), entry.getValue()))
+      .collect(Collectors.joining("\n"));
+    configMap.putDataItem(podInfo.getLabelsFile(), labelsStr);
+    if (podInfo.getName() != null) {
+      configMap.putDataItem(podInfo.getNameFile(), podInfo.getName());
+    }
+    LOG.trace("Creating config-map {} in Kubernetes", resourceMeta.getName());
+    configMap = api.createNamespacedConfigMap(kubeNamespace, configMap, "true");
+    LOG.trace("Created config-map {} in Kubernetes", resourceMeta.getName());
+    return configMap;
```

Comment:
Set owner references to the config map as well.

Reply:
they are set. Every resource here will share the same meta object, just will have a different kind.